### DISCUSSION TITLE
file.recurse: use different renderers in individual files

### DISCUSSION THREAD

**Discussion by xorinzor** [saltstack/salt#65335] [Q&A] [answered]
I have a directory on my saltmaster containing multiple files that need to be added to the target host. Some of which should be copied as-is, but others should be rendered using `jinja`.

The way I currently solve this is:
- I add a `.j2` extension to all my jinja files
- In my SLS file I first copy all non-jinja files as-is via a `file.recurse` state which excludes all jinja files using the `exclude_pat: *.j2` attribute.
- Next, for every jinja file, I have to add an `file.managed` state because I need to remove all `j2` extensions (AFAIK I can't do this in bulk via file.recurse).

In the hopes of a cleaner solution, I tried adding `#!jinja|yaml` at the top of my managed file in the hopes that SaltStack would pick up on this within `file.recurse`, but it unfortunately just copied the file as-is to the target host as it seems this `shebang` method only works for SLS files.

Is there any way to define within the file itself how the `file.recurse` state should render the file? Ie: if it should copy as-is, use `jinja` etc.? Or is there an alternative better / cleaner solution to this problem?

I could alternatively use different subdirectories for each type of file (as-is, jinja, etc.) but this feels like I'm just creating different workarounds.

**Comment by MartinEmrich** [comment, marked as answer]
You could render all files as jinja, but in the files not to be templated, enclose everything in

````
{%- raw -%}
  ...
{%- endraw -%}
````
if there's jinja-like markup in there.
